File: examples/java/src/main/java/runar/examples/post-quantum-wots/PostQuantumWOTS.runar.java
package runar.examples.postquantumwots;

import runar.lang.SmartContract;
import runar.lang.annotations.Public;
import runar.lang.annotations.Readonly;
import runar.lang.types.ByteString;

import static runar.lang.Builtins.assertThat;
import static runar.lang.Builtins.verifyWOTS;

/**
 * PostQuantumWOTS -- minimal conformance fixture exercising the
 * one-time WOTS+ post-quantum signature verification primitive.
 */
class PostQuantumWOTS extends SmartContract {

    @Readonly ByteString pubkey;

    PostQuantumWOTS(ByteString pubkey) {
        super(pubkey);
        this.pubkey = pubkey;
    }

    @Public
    void spend(ByteString msg, ByteString sig) {
        assertThat(verifyWOTS(msg, sig, this.pubkey));
    }
}

File: examples/ts/post-quantum-wots/PostQuantumWOTS.runar.ts
import { SmartContract, assert, verifyWOTS } from 'runar-lang';
import type { ByteString } from 'runar-lang';

class PostQuantumWOTS extends SmartContract {
  readonly pubkey: ByteString;

  constructor(pubkey: ByteString) {
    super(pubkey);
    this.pubkey = pubkey;
  }

  public spend(msg: ByteString, sig: ByteString) {
    assert(verifyWOTS(msg, sig, this.pubkey));
  }
}
